src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Helpers.cs

```diff
@@ -5,15 +5,118 @@
 using System.Buffers.Text;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 
 namespace System.Text.Json
 {
     public sealed partial class Utf8JsonWriter
     {
+        private const byte LengthMask =         0b000_000_11;
+        private const byte EncodingMask =       0b000_111_00;
+
+        private const byte Utf8EncodingFlag =   0b000_001_00;
+        private const byte Utf16EncodingFlag =  0b000_010_00;
+
+        private bool TryGetPartialUtf8CodePoint(out ReadOnlySpan<byte> codePointBytes)
+        {
+            ReadOnlySpan<byte> partialCodePointBytes = PartialCodePointRaw;
+            Debug.Assert(partialCodePointBytes.Length == 4);
+
+            if ((partialCodePointBytes[3] & Utf8EncodingFlag) == 0)
```

Comment:
Nit: storing the flags within the same buffer that also contains partial code point seems a bit strange to me. Couldn't it be stored in a separate field e.g. `byte _partialCodePointFlags;` to better emphasize the distinction from data?

Reply:
Updated. For framework it adds 1 byte to the struct that would have otherwise been in the heap array. However, due to packing I think it won't actually add to the size of the struct (there are 3 other byte/bool fields).

Reply:
Sorry, what struct? Can't it just be a field on the `Utf8JsonWriter` class?

Reply:
Yep, I updated it. I was mixing up the two platforms.